package/MDAnalysis/topology/MMCIFParser.py

```diff
@@ -0,0 +1,178 @@
+    Elements,
+    FormalCharges,
+    ICodes,
+    Masses,
+    Occupancies,
+    RecordTypes,
+    Resids,
+    Resnames,
+    Resnums,
+    Segids,
+    Tempfactors,
+)
+from .base import TopologyReaderBase
+
+
+def _into_idx(arr: list[int]) -> list[int]:
+    return [idx for idx, (_, group) in enumerate(itertools.groupby(arr)) for _ in group]
+
+
+class MMCIFParser(TopologyReaderBase):
+    format = "MMCIF"
+
+    def parse(self, **kwargs):
+        """Read the file and return the structure.
+
+        Returns
+        -------
+        MDAnalysis Topology object
+        """
+        structure = gemmi.read_structure(self.filename)
+
+        if len(structure) > 1:
+            warnings.warn(
+                "MMCIF model {self.filename} contains {len(model)=} different models, "
+                "but only the first one will be used to assign the topology"
+            )
+        model = structure[0]
+
+        # atom properties
+        (
+            altlocs,  # at.altloc
+            serials,  # at.serial
+            names,  # at.name
+            atomtypes,  # at.name
+            # ------------------
+            chainids,  # chain.name
+            elements,  # at.element.name
+            formalcharges,  # at.charge
+            weights,  # at.element.weight
+            # ------------------
+            occupancies,  # at.occ
+            record_types,  # res.het_flag
+            tempfactors,  # at.b_iso
+            residx,  # _into_idx(res.seqid.num)
+        ) = map(  # this is probably not pretty, but it's efficient -- one loop over the mmcif
+            np.array,
+            list(
+                zip(
+                    *[
+                        (
+                            at.altloc,  # altlocs
+                            at.serial,  # serials
+                            at.name,  # names
+                            at.name,  # atomtypes
+                            # ------------------
+                            chain.name,  # chainids
+                            at.element.name,  # elements
+                            at.charge,  # formalcharges
+                            at.element.weight,  # weights
+                            # ------------------
+                            at.occ,  # occupancies
+                            res.het_flag,  # record_types
+                            at.b_iso,  # tempfactors
+                            res.seqid.num,  # residx, later translated into continious repr
+                        )
+                        for chain in model
+                        for res in chain
+                        for at in res
+                    ]
+                )
+            ),
+        )
+
+        (
+            icodes,  # res.seqid.icode
+            resids,  # res.seqid.num
+            resnames,  # res.name
+            segidx,  # chain.name
+            resnums,
+        ) = map(
+            np.array,
+            list(
+                zip(
+                    *[
```

Comment:
I'm struggling to follow the logic here, a comment breaking down what this double nested loop iteration into a zip is doing would be nice